src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs

```diff
@@ -8,98 +8,164 @@
+                  supports.ToDictionary(r => r.Name))
         {
         }
 
-        private RuntimeGraph(Dictionary<string, RuntimeDescription> runtimes, Dictionary<string, CompatibilityProfile> supports)
+        private RuntimeGraph(IReadOnlyDictionary<string, RuntimeDescription> runtimes, IReadOnlyDictionary<string, CompatibilityProfile> supports)
         {
-            Runtimes = new ReadOnlyDictionary<string, RuntimeDescription>(runtimes);
-            Supports = new ReadOnlyDictionary<string, CompatibilityProfile>(supports);
+            Runtimes = runtimes;
+            Supports = supports;
+
+            if (Runtimes.Count != 0 || Supports.Count != 0)
+            {
+                _areCompatible = new();
+                _expandCache = new(StringComparer.Ordinal);
+                _dependencyCache = new();
+            }
         }
 
+        internal bool IsEmpty => Runtimes.Count == 0 && Supports.Count == 0;
+
         public RuntimeGraph Clone()
         {
-            return new RuntimeGraph(Runtimes.Values.Select(r => r.Clone()), Supports.Values.Select(s => s.Clone()));
+            if (IsEmpty)
+            {
+                return this;
+            }
+
+            return new RuntimeGraph(
+                runtimes: Clone(Runtimes, r => r.Clone()),
+                supports: Clone(Supports, s => s.Clone()));
+
+            static IReadOnlyDictionary<string, T> Clone<T>(IReadOnlyDictionary<string, T> source, Func<T, T> cloneFunc)
+            {
+                if (source.Count == 0)
+                {
+                    // No need to allocate anything
+                    return source;
+                }
+
+                Dictionary<string, T> clone = new(source.Count);
+
+                // TODO append .NoAllocEnumerate() here once https://github.com/NuGet/NuGet.Client/pull/5246 is available
+                foreach (var pair in source)
+                {
+                    clone[pair.Key] = cloneFunc(pair.Value);
+                }
+
+                return clone;
+            }
         }
 
         /// <summary>
-        /// Merges the content of the other runtime graph in to this runtime graph
+        /// Merges the content of two runtime graphs and returns the combined graph.
         /// </summary>
         /// <param name="other">The other graph to merge in to this graph</param>
         public static RuntimeGraph Merge(RuntimeGraph left, RuntimeGraph right)
         {
-            var runtimes = new Dictionary<string, RuntimeDescription>();
-            foreach (var runtime in left.Runtimes.Values)
+            if (left.IsEmpty)
+            {
+                return right;
+            }
+
+            if (right.IsEmpty)
             {
-                runtimes[runtime.RuntimeIdentifier] = runtime.Clone();
+                return left;
+            }
+
+            var runtimes = new Dictionary<string, RuntimeDescription>(capacity: left.Runtimes.Count + right.Runtimes.Count);
```

Comment:
Specify the size on the dictionary to avoid resizing. In this case we know the exact size required, so use it.